File: dailyQuestion/2020-06/06-18/python/solution_item.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def recoverFromPreorder(S: str) -> TreeNode:
    if not S:
        return None
    stack, pos = list(), 0
    while pos < len(S):
        level = 0
        while S[pos] == "-":
            pos += 1
            level += 1
        value = 0
        while pos < len(S) and S[pos].isdigit():
            value = value * 10 + int(S[pos])
            pos += 1
        node = TreeNode(value)
        if level == len(stack):
            if stack:
                stack[-1].left = node
        else:
            stack = stack[:level]
            stack[-1].right = node
        stack.append(node)
    return stack[0]


if __name__ == "__main__":
    S = "1-2--3--4-5--6--7"
    result = recoverFromPreorder(S)
    print(result)

File: dailyQuestion/2020-06/06-18/golang/solution_item.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	S := "1-2--3--4-5--6--7"
	result := recoverFromPreorder(S)
	fmt.Println(result)
}

func recoverFromPreorder(S string) *TreeNode {
	if len(S) == 0 {
		return nil
	}
	stack := []*TreeNode{}
	pos := 0
	for pos < len(S) {
		level := 0
		for S[pos] == '-' {
			level++
			pos++
		}
		value := 0
		for pos < len(S) && (S[pos] >= '0' && S[pos] <= '9') {
			value = value*10 + int(S[pos]-'0')
			pos++
		}
		node := &TreeNode{Val: value}
		if level == len(stack) {
			if len(stack) != 0 {
				stack[len(stack)-1].Left = node
			}
		} else {
			stack = stack[:level]
			stack[level-1].Right = node
		}

		stack = append(stack, node)
	}
	return stack[0]
}
